File: frontend/src/types/explanation.ts
export interface FeatureContributionRecord {
  feature_name: string;
  feature_value: any;
  importance_score: number;
  shap_value: number;
  direction: "INCREASES_RISK" | "DECREASES_RISK" | "NEUTRAL";
  description: string;
}

export interface ModelContributionRecord {
  model_id: string;
  model_name: string;
  model_type: string;
  weight: number;
  risk_score: number;
  contribution_score: number;
  status: string;
}

export interface BusinessRuleContributionRecord {
  rule_id: string;
  rule_name: string;
  rule_category: string;
  severity: string;
  action: string;
  impact_score: number;
  description: string;
}

export interface RecommendationRecord {
  action_type: string;
  title: string;
  rationale: string;
  priority: string;
  metadata: Record<string, any>;
}

export interface ExplanationRecord {
  id: string;
  explanation_id: string;
  decision_id: string;
  transaction_id: string;
  merchant_id?: string;
  customer_id?: string;
  composite_risk_score: number;
  confidence_score: number;
  primary_reason: string;
  feature_contributions: FeatureContributionRecord[];
  model_contributions: ModelContributionRecord[];
  rule_contributions: BusinessRuleContributionRecord[];
  recommendations: RecommendationRecord[];
  audit_info: {
    audit_hash: string;
    engine_version: string;
    compliance_standard: string;
    audited_at: string;
    verifiable: boolean;
  };
  created_at: string;
  updated_at: string;
}

export interface PaginatedExplanations {
  items: ExplanationRecord[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface ExplanationFilterParams {
  search?: string;
  page?: number;
  size?: number;
}

File: backend/app/schemas/explanation.py
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class FeatureContributionSchema(BaseModel):
    feature_name: str
    feature_value: Any
    importance_score: float
    shap_value: float
    direction: str
    description: str


class ModelContributionSchema(BaseModel):
    model_id: str
    model_name: str
    model_type: str
    weight: float
    risk_score: float
    contribution_score: float
    status: str


class BusinessRuleContributionSchema(BaseModel):
    rule_id: str
    rule_name: str
    rule_category: str
    severity: str
    action: str
    impact_score: float
    description: str


class RecommendationSchema(BaseModel):
    action_type: str
    title: str
    rationale: str
    priority: str
    metadata: Dict[str, Any] = Field(default_factory=dict)


class ExplanationGenerateRequest(BaseModel):
    decision_id: str = Field(..., example="DEC-8B13D12B", description="Target Decision ID to explain")


class ExplanationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    explanation_id: str
    decision_id: str
    transaction_id: str
    merchant_id: Optional[uuid.UUID] = None
    customer_id: Optional[str] = None
    composite_risk_score: float
    confidence_score: float
    primary_reason: str
    feature_contributions: List[FeatureContributionSchema]
    model_contributions: List[ModelContributionSchema]
    rule_contributions: List[BusinessRuleContributionSchema]
    recommendations: List[RecommendationSchema]
    audit_info: Dict[str, Any]
    created_at: datetime
    updated_at: datetime


class PaginatedExplanationsResponse(BaseModel):
    items: List[ExplanationResponse]
    total: int
    page: int
    size: int
    pages: int
